File: Python/Ad.py
from BaseAdvertising import BaseAdvertising


class Ad(BaseAdvertising):
    """represents an ad that is owned by an advertiser"""
    id_counter: int = 1

    def __init__(self, title, img_url, link, advertiser):
        super().__init__()
        self.__id = Ad.id_counter
        self.__title = title 
        self.__img_url = img_url
        self.__link = link 
        self.__advertiser = advertiser
        Ad.id_counter += 1

    def get_title(self):
        return self.__title

    def set_title(self, title):
        self.__title = title

    def get_img_url(self):
        return self.__img_url
    
    def set_img_url(self, img):
        self.__img_url = img

    def get_link(self):
        return self.__link

    def set_link(self, link):
        self.__link = link 

    def set_advertiser(self, advertiser):
        self.__advertiser = advertiser

    def inc_clicks(self):
        super().inc_clicks()
        self.__advertiser.inc_clicks()

    def inc_views(self):
        super().inc_views()
        self.__advertiser.inc_views()

    def describe_me(self):
        return self.__doc__

File: Cpp/src/Ad.cpp
#include <string>
#include "Ad.hpp"
using namespace std;

int Ad::id_counter;

Ad::Ad(std::string title, std::string img_url, std::string link, Advertiser *advertiser){
    this->title = title;
    this->img_url = img_url;
    this->link = link;
    this->advertiser = advertiser;
    this->id = ++Ad::id_counter;
}

std::string Ad::get_title(){
    return title;
}

void Ad::set_title(std::string title){
    this->title = title;
}

std::string Ad::get_img_url(){
    return img_url;
}

void Ad::set_img_url(std::string img_url){
    this->img_url = img_url;
}

std::string Ad::get_link(){
    return link;
}

void Ad::set_link(std::string link){
    this->link = link;
}

void Ad::set_advertiser(Advertiser *advertiser){
    this->advertiser = advertiser;
}

void Ad::inc_clicks(){
    advertiser->inc_clicks();
    BaseAdvertising::inc_clicks();
}

void Ad::inc_views(){
    advertiser->inc_views();
    BaseAdvertising::inc_views();
}

std::string Ad::describe_me(){
    return "Ad class represents an ad that is owned by an advertiser";
}
